core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -55,10 +55,26 @@ trait RTS {
   def defaultHandler: List[Throwable] => IO[Nothing, Unit] =
     (ts: List[Throwable]) => IO.sync(ts.foreach(_.printStackTrace()))
 
+  /**
+   * The minimum number of threads maintained in the thread pool.
+   */
+  val corePoolSize = 2
+
+  /**
+   * The maximum number of threads allowed in the thread pool.
+   */
+  val maximumPoolSize = Runtime.getRuntime().availableProcessors().max(corePoolSize)
+
   /**
    * The main thread pool used for executing fibers.
    */
-  val threadPool = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors().max(2))
+  val threadPool = new ThreadPoolExecutor(
+    corePoolSize,
+    maximumPoolSize,
+    60000,
+    TimeUnit.MILLISECONDS,
+    new LinkedBlockingQueue[Runnable]()
+  )
```

Comment:
Using constants for `corePoolSize` and `maximumPoolSize` is not reliable; because of the order of construction. So in my opinion we should not do that.

Instead, here's what I'd propose for the default thread pool:

 - `0` as the minimum number of threads 
 - `Int.MaxValue` as the maximum number of threads 
 - 100 milliseconds as the idle timeout
 - unbounded blocking queue as already being done

Reply:
In one of @kaishh comment to this PR, it was pointed out that the consequence of using an unbounded work queue is that the pool will never spawn new threads since an unbounded queue will never be full. We can use a `SynchronousQueue` (effectively an internal capacity of 0) or an `ArrayBlockingQueue` to always spawn a new thread if none are idle when a task is submitted.

It was also my understanding that this pool was intended for CPU-bound workloads. Do we want to instead have, as a sensible default, a cached pool similar to `ExecutionContext.Implicits.global`, which users can override to better fit their workloads? I could imagine scenarios where a lot of overhead is incurred in creating and reaping threads so fast.

Reply:
> Int.MaxValue as the maximum number of threads

The machine will die on several thousands.

> 100 milliseconds as the idle timeout

This is too low, it has to be way bigger than potential overhead of thread spawning.

> unbounded blocking queue as already being done

Not a sane option as well. It's a way to OOM (don't forget about publisher/consumer speed disbalance problem). Also unbounded queue will prevent the pool from growing.

Please look at my comments above as well.

Reply:
> It's a way to OOM

That's more or less ok, I think. IMHO ZIO model shouldn't set limits on fiber count beyond just VM memory capacity, at least in `App` RTS.

> Also unbounded queue will prevent the pool from growing.

If we assume that the pool is for CPU-bound tasks, the pool probably shouldn't grow at all, except specifically when a task wrapped  **scala.concurrent.blocking** is spawned – that's the way `global` pool and monix manage it – OTOH, ZIO probably should schedule those on a separate unbounded pool right away.

Reply:
Good point about unbounded queue. However, 100 milliseconds is way longer than it takes to start a new thread. I'd like threads to die quickly so users can expect a speedy exit when running their apps or testing things from the REPL.

```scala
def zio(threadFactory: ThreadFactory): ExecutorService = {
    val cores = Runtime.getRuntime.availableProcessors.max(2)
    val expectedFibersPerRequest = 2 * 3

    val executor = new ThreadPoolExecutor(
      cores
      , cores * cores
      , 60
      , TimeUnit.SECONDS
      , new LinkedBlockingQueue[Runnable](cores * expectedFibersPerRequest)
      , threadFactory
    )
    executor
  }
```

Aside from idle time, which I'd still prefer to keep to 100 milliseconds, I think this looks great, although we will have to pick a constant multiplier for the size of the queue.